### ISSUE TITLE
META: glibc baseline in conda-forge - moving to 2.28 / alma8

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/conda-forge.github.io#2383]
Since we've separated the baselines between docker_images and the sysroot (and the docker_image [bump](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6626) has now happened), this question will inevitably come up, so we might as well have an issue to reference and discuss.

# Note: this is likely still years away!

With that said, here's the background. The timeline of changes for our images and sysroot baseline is as follows

| when | sysroot | docker_image | comment |
|-|-|-|-|
| up to Nov. 2021 | 2.12 (cos6) | cos6 | matching distro version, until discussion [here](https://github.com/conda-forge/conda-forge.github.io/issues/1436) eventually lead to split |
| Dec 2021 | 2.12 (cos6) | cos7 | https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/2241 | 
| July 2024 | 2.17 (cos7) | cos7 | This needed a lot of preparation (mainly rolling out [`stdlib`](https://github.com/conda-forge/conda-forge.github.io/issues/2102)); finally done [here](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6070) | 
| Nov 2024 | 2.17 (cos7) | alma9 | Due to how long things got held up (for stdlib & [CDTs](https://github.com/conda-forge/cdt-builds/issues/66)), we actually<br/>ended up leapfrogging [alma8](https://github.com/conda-forge/conda-forge.github.io/issues/1941) and went directly to [alma9](https://github.com/conda-forge/conda-forge.github.io/issues/2342) | 
| ??? | ? | alma10 | Will likely release in 2025, but it's unlikely to really be necessary</br>as infrastructure for packages until at least 2-3 years after |
| ??? | 2.28 (alma8) | ? | The topic of this issue 🙃  |

The reason we are able to separate these two things is because we can distinguish between what we compile against (i.e. the sysroot that we package and pin to a specific version), from the actual glibc version that's present in the image. Some more explanation is in this [announcement](https://conda-forge.org/news/2024/11/22/new-images/).

Part of the discussion in #1436 and why it took so long to move on, is that conda-forge has some large users (e.g. academic institutions) that move glacially w.r.t. infrastructure updates, and we try to support those use-cases as long as we feasibly can (not least because some core members are involved there). That meant we [went](https://github.com/conda-forge/conda-forge.github.io/pull/1980) all the way to the very end-of-life (even with the paid extension that's not actually relevant for us!) of `cos6` before dropping it. This was not actually a causal relationship, but just a reflection of how much we've tried to stay on cos6 as long as feasible.

For the move from `cos7` to `alma8`, the situation will very likely repeat, in that we want to be as defensive as possible with upgrading the baseline, while enabling individual feedstocks to pull in newer baseline (using `c_stdlib_version`) as necessary. A good reference is https://mayeut.github.io/manylinux-timeline/, which provides an overview of the consumer glibc version across all downloads from PyPI[^1]. As of Nov. 2024, around 85% of PyPI downloaders on linux have `__glibc>=2.28`, and this number rises to 94% if we're not counting people on python versions that are already EOL.

[^1]: I'd really like to have an equivalent for conda-forge, but that's another story

Note however that we didn't drop glibc 2.12 until it was below 0.1% of downloads (at least via the proxy of the PyPI downloads). Perhaps that's a little extreme (and also owed due to the stdlib-work that took a while to put in place), but in any case, it will likely have to be a _very_ low percentage of users or there'd have to be very strong technical constrains that force our hand before we move on.

In short: enjoy the new alma-based infrastructure, move the glibc baseline for your feedstock if it cannot be avoided, but the global baseline will likely still take quite a while to move.

**Comment by h-vetinari** [comment]
Even though it's not 100% representative of conda users, a good proxy for glibc version distribution in the ecosystem is https://mayeut.github.io/manylinux-timeline. Since the data eventually drops off after 2 years, I wanted to record the following historical relationship. The last day where glibc 2.12-2.16 had a detectable blip of usage was in Oct. 2023 (after that it's 0.0%).

We ended up moving to glibc 2.17 in July 2024, well after usage (as approximated by PyPI downloads) had dropped below 1 in 1000 users (also, this is the broadest measure, including downloads on EOL python versions, which we don't serve anymore anyway).

![Image](https://github.com/user-attachments/assets/fbd71e88-4d24-414a-b6c0-76ecaa3f6cef)

Part of being so late for dropping 2.12 was certainly due to not having `{{ stdlib("c") }}` in place, but let's assume for now that we want to stay similarly conservative for dropping 2.17 (barring strong technical reasons that force us to move earlier). Currently, the number of users on glibc 2.17-2.27 (i.e. the versions we'd drop if we moved to alma8 / 2.28) represent ~6.5% of users on python versions that aren't EOL yet, and ~10.7% when including EOL pythons. 

![Image](https://github.com/user-attachments/assets/ef11cd48-c19a-4f2d-b642-14d1fe7a09fc)

**Comment by daneos** [comment]
Is there some rough timeline for moving the baseline? As I understand this is dependent on number of glibc 2.17-2.27 users but is there some rough expectation of when this might happen?

**Comment by h-vetinari** [comment]
There's no timeline. Basically the discussion will start once some core packages in the ecosystem require `glibc >=2.28` (i.e. if a substantial portion of the packages we build have some transitive dependency that requires it).

It's theoretically possible that we move the baseline because we see that user numbers on `glibc <2.28` are basically non-existent (that would need better data collection along the lines of https://github.com/conda/infrastructure/issues/1018, c.f. also https://github.com/anaconda/anaconda-package-data/issues/64), but historically there's been strong opposition to dropping support for anything unless there was a strong technical reason to do so.

**Comment by h-vetinari** [comment]
Worth noting that Anaconda's channels began transitioning to a baseline of glibc 2.28 with https://github.com/AnacondaRecipes/aggregate/commit/8abbad99aa658b56e94b3f56266da7613fb7ade2 this summer, and now set homogeneously for all new linux builds as of two months ago: https://github.com/AnacondaRecipes/aggregate/commit/e9bb747ecc4aaf7b47099c9a112fcba2d8b5d246

**Comment by h-vetinari** [comment]
Pretty big slowdown (-25% for fitting) in scikit-learn was [tracked down](https://github.com/conda-forge/ctng-compiler-activation-feedstock/issues/108#issuecomment-5541436973) to an old glibc. Also worth noting that RHEL 7 is [EOL](https://endoflife.date/rhel) since 2 years. It may slowly become worth discussing a timeline for bumping to 2.28.

Relevant sidenote: Google projects (abseil etc.) have recently moved their baseline from 2.27 to 2.34 already (see [here](https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4075#issuecomment-5484748821)), so in any case we'll have to keep handling this aspect in the respective migrators.

**Comment by hmaarrfk** [comment]
Just a note that Ubuntu 20.04 is glibc 2.31.

**Comment by ax3l** [comment]
Small note: glibc adds many vectorized math functions in 2.35+.
We currently bracket quite coarsely by building 3.34 + 2.39 (due to RHEL builds) and thus miss proper performance for vectorized code on Ubuntu 22.04 (glibc 2.35).

**Comment by h-vetinari** [comment]
> glibc adds many vectorized math functions in 2.35+.

Affected feedstocks (which benefit specifically from these) may be able to selectively add an additional build against 2.39. But I don't want to get sidetracked with discussing the benefits of RHEL 10, when even just raising the glibc baseline to RHEL 8 will likely face some resistance.

**Comment by mikekuzak** [comment]
RedHat 8 is EOL and it's derivatives are EOL in 2029. Meaning you can't just move to anything higher than 2.28. Any OS with a higher glib is compatible with a lower version, but yiu know that. 2.28 is the way, as a next step.

**Comment by anderslindho** [comment]
FWIW many scientific facilities and labs are - to various degrees - stuck on or must maintain RHEL 7 (e.g. CentOS 7) installations.

**Comment by h-vetinari** [comment]
> FWIW many scientific facilities and labs are - to various degrees - stuck on or must maintain RHEL 7 (e.g. CentOS 7) installations.

Can you give some examples? Conda-forge users in such institutions should ideally be participating in these discussions.

**Comment by anderslindho** [comment]
I can give one example: the European Spallation Source (ESS). I know many more sites - like ITER, ANL, BNL, PSI - are required to maintain old installs, but I don't know whether they use conda-forge.

**Comment by ax3l** [comment]
In my experience, these are usually the sites that already delegate to their users to build their own stack for everything or advise to run in containers... And they can continue to use all existing pre 2026 CF packages without any issues if we were to update. Reproducibility is still there for people that like to run EOL installs.

**Comment by anderslindho** [comment]
Maybe true for some. But ESS' control system uses and is built on top of conda-forge, and many of their ~1 500 control system servers are still on CentOS 7 with no immediate opportunity to upgrade (even though the transition has started and is underway).

**Comment by h-vetinari** [comment]
> I know many more sites - like ITER, ANL, BNL, PSI - are required to maintain old installs

Is there a person or institution actually _requiring_ them to (and if so, why)? Perhaps I'm misunderstanding, but your formulation sounded as if it's something beyond a simple need stemming from having productive machines that are still on such old distro versions.

> But ESS' control system uses and is built on top of conda-forge

That's cool to know!

> and many of their ~1 500 control system servers are still on CentOS 7 with no immediate opportunity to upgrade (even though the transition has started and is underway). 

Is there a rough timeline for this transition?

**Comment by anderslindho** [comment]
> Is there a person or institution actually requiring them to (and if so, why)? Perhaps I'm misunderstanding, but your formulation sounded as if it's something beyond a simple need stemming from having productive machines that are still on such old distro versions.

Not really, no, outside of resource availability (for the updates) and "don't touch it if it works". Sorry if initially unclear, but these old distro machines will generally fall into one out of two categories: black box, stable in prod, won't have any updates from conda-forge (nor elsewhere); semi black box, stable in prod, gets small updates but difficult to convince stakeholders that a full distribution update is wanted.

> Is there a rough timeline for this transition?

Not really, no. The target was yesterday/ASAP but this sort of thing takes time and unfortunately falls far lower in prio than development. I think full migration within 2 years is feasible. I think loss of conda-forge support for updates in a (far) shorter time than that is reasonable. For full disclosure, I left them a week ago, and am not speaking in any official capacity.

**Comment by h-vetinari** [comment]
Thanks for the context. Perhaps you still have contacts there and can ping your old boss/sysadmin (or whoever's somewhat responsible for this transition) to participate in this discussion?

We do try to keep things running as long as possible where we can, but we can't do it indefinitely either - the performance issue in scikit-learn is a good illustration of the "cost" that our users pay from us prioritising compatibility, and even though individual cases can be worked-around (once someone correctly rootcauses the issue, which itself is also far from a given...), the accumulated tech debt of doing that all over the place becomes too great at some point to justify a handful of remaining users (_especially_ if those are essentially invisible to us - hence why I'm suggesting to at least establish that line of communication).

**Comment by anderslindho** [comment]
Of course. Summoning @lucmaga.

**Comment by lucmaga** [comment]
> Not really, no. The target was yesterday/ASAP but this sort of thing takes time and unfortunately falls far lower in prio than development. I think full migration within 2 years is feasible. I think loss of conda-forge support for updates in a (far) shorter time than that is reasonable. For full disclosure, I left them a week ago, and am not speaking in any official capacity.

That is mostly correct. I want to be more optimistic saying we could handle most of critical systems before that, but still can foreseen at least 1 year still.

> Maybe true for some. But ESS' control system uses and is built on top of conda-forge, and many of their ~1 500 control system servers are still on CentOS 7 with no immediate opportunity to upgrade (even though the transition has started and is underway).

@anderslindho  I think, 1500 control system hosts running Centos might be a exaggeration at this point. What I see here is ~500, still most of them are critical.

**Comment by anderslindho** [comment]
I never said 1 500 servers on CentOS, Lucas 😄 

I left a repo that fetches you up to date stats - it's under the icshwi namespace and is called something like centos-migration or ubuntu-migration.

**Comment by lucmaga** [comment]
I misinterpreted your `~1 500` notation then.
